### PR TITLE
mesa gallium

### PR DESCRIPTION
**Pull request by luigi-calori**
Try to compile mesa with gallium drivers to support multithreading opengl emulation on nodes without GPU

### CODE REVIEW

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 1]
```diff
@@ -30,16 +30,29 @@ class Mesa(AutotoolsPackage):
     specification - a system for rendering interactive 3D graphics."""
 
     homepage = "http://www.mesa3d.org"
-    url      = "http://ftp.iij.ad.jp/pub/X11/x.org/pub/mesa/12.0.3/mesa-12.0.3.tar.gz"
+    base_url = "ftp://ftp.freedesktop.org/pub/mesa/"
+    url      = base_url+"12.0.3/mesa-12.0.3.tar.gz"
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: @adamjstewart do we need `url` if there is `url_for_version`?

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 2]
```diff
@@ -30,16 +30,29 @@ class Mesa(AutotoolsPackage):
...
 
     version('12.0.3', '60c5f9897ddc38b46f8144c7366e84ad')
-
+    version('13.0.2', 'cb828f9f68e0cd6c7ef30b9c67aa7cf2')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: please sort versions in descending order, i.e. `13.0.2` first, then `13.0.1`, etc...

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 3]
```diff
@@ -30,16 +30,29 @@ class Mesa(AutotoolsPackage):
...
     # General dependencies
     depends_on('python@2.6.4:')
     depends_on('py-mako@0.3.4:', type=('build', 'run'))
     depends_on('flex@2.5.35:', type='build')
     depends_on('bison@2.4.1:', type='build')
 
+    variant('gallium', default=False, description="compile with gallium llvm sw rendering")
+    variant('drm', default=True, description="compile with drm")
+
+    #depends_on("llvm@3.0", when='@8.0.5~gallium')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: this can probably be removed, same below.

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 4]
```diff
@@ -50,12 +63,81 @@ class Mesa(AutotoolsPackage):
...
     depends_on('presentproto@1.0:', type='build')
     depends_on('pkg-config@0.9.0:', type='build')
+    depends_on('binutils', type='build')
 
     # TODO: Add package for systemd, provides libudev
     # Using the system package manager to install systemd didn't work for me
+
+    def url_for_version(self, version):
+        """Handle mesa version-based custom URLs."""
+        latest_mver = '11'
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: i would probably use it directly in `Version('11')` and add a comment above the `if` condition
```
# if it's not the latest version:
```

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 5]
```diff
@@ -50,12 +63,81 @@ class Mesa(AutotoolsPackage):
...
+
+        if version < Version(latest_mver):
+            base = Mesa.base_url + "/older-versions/%s.x" % (version.up_to(1))
+        else:
+            base = Mesa.base_url 
+        if version > Version('10.4'):
+            name = 'mesa'
+        else:
+            name = 'MesaLib'
+        ret = base + "/%s/%s-%s.tar.gz" % (version, name, version)
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: just return directly

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 6]
```diff
@@ -50,12 +63,81 @@ class Mesa(AutotoolsPackage):
...
+
+#    def install(self, spec, prefix):
+#        configure('--prefix={0}'.format(prefix))
+#        self.check_variants(
+#            var='+gallium',
+#            ver='@:9.2.2',
+#            ref='http://www.paraview.org/Wiki/ParaView/ParaView_And_Mesa_3D')
+
+    def configure_args(self):
+        self.check_variants(
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: with the current `develop`, the neat way of writing this is
```
conflicts('+galium', when='@:9.2.2')
```
then you can remove `check_variants`.

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 7]
```diff
@@ -50,12 +63,81 @@ class Mesa(AutotoolsPackage):
...
+        return ret
+
+    def check_variants(self, var='', ver='', ref=''):
+        if var and ver:
+            error = "to be safe avoid '{variant}' with version '{version}' see:\n {reference}" # noqa
+            if var in self.spec and self.spec.satisfies(ver):
+                raise RuntimeError(
+                    error.format(variant=var, version=ver, reference=ref))
+
+#    def install(self, spec, prefix):
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: remove if not needed.

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 8]
```diff
@@ -50,12 +63,81 @@ class Mesa(AutotoolsPackage):
...
+                                 join_path(self.prefix, 'lib'))
+            run_env.set('GALLIUM_DRIVER', 'swr')
+
+    def setup_dependent_environment(self, spack_env, run_env, dspec):
+            run_env.prepend_path('LD_LIBRARY_PATH',
+                                 join_path(self.prefix, 'lib'))
+            run_env.set('GALLIUM_DRIVER', 'swr')
+
+      
+#        options = ['-prefix=%s' % prefix]
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: remove if not needed.

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 6 above)
Comment (citibeth) [inline var/spack/repos/builtin/packages/mesa/package.py]: Yes definitely.  I didn't know what `check_variants()` did anyway, and @davydden's suggestion makes the intent more clear.  However... then you don't get the `ref=`, would that negatively affect functionality?

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 9]
```diff
@@ -30,16 +30,29 @@ class Mesa(AutotoolsPackage):
...
+    version('13.0.0', '7205edb90d0396dc26d049fa495f6fd1')
+   
     # General dependencies
     depends_on('python@2.6.4:')
     depends_on('py-mako@0.3.4:', type=('build', 'run'))
     depends_on('flex@2.5.35:', type='build')
     depends_on('bison@2.4.1:', type='build')
 
+    variant('gallium', default=False, description="compile with gallium llvm sw rendering")
+    variant('drm', default=True, description="compile with drm")
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: this variant is not used!

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 6 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: the link here is a general description, one can just keep it as a comment above
```
# see http://www.paraview.org/Wiki/ParaView/ParaView_And_Mesa_3D
conflicts('+galium', when='@:9.2.2')
```
or we need to ask @alalazo to allow custom messages in `confligts()` statement :smile:

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: You do if you want `spack versions` and `spack checksum` to work.

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 10]
```diff
@@ -50,12 +63,81 @@ class Mesa(AutotoolsPackage):
...
     depends_on('libxdamage')
     depends_on('libxfixes')
+    depends_on('expat')
 
     depends_on('glproto@1.4.14:', type='build')
     depends_on('dri2proto@2.6:', type='build')
     depends_on('dri3proto@1.0:', type='build')
     depends_on('presentproto@1.0:', type='build')
     depends_on('pkg-config@0.9.0:', type='build')
+    depends_on('binutils', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: We generally avoid adding binutils dependencies because technically every package in Spack depends on ld, much like every package depends on curl to fetch it. It's more of a Spack dependency than a Mesa dependency.

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 6 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: another big difference is that `conflicts()` will fail during conretization whereas those checks will fail during installation of this package. We do want to stick with the former!

**Comment by davydden** [comment]
@luigi-calori could please address the comments and hopefully it will be merged soon.

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 8 above)
Comment (luigi-calori) [inline var/spack/repos/builtin/packages/mesa/package.py]: If I remember well, this is needed to activate the gallium driver that  is much faster than the others

Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: @luigi-calori are you sure you intended to leave reply to this comment? Does not seem to be related...

Comment (luigi-calori) [inline var/spack/repos/builtin/packages/mesa/package.py]: Sorry, newbie of github code diff commenting, I thought you were suggesting to remove one of the run_env.set('GALLIUM_DRIVER', 'swr')
This is somehow needed by the environment that depends on mesa to select the driver
Of course the trailing commented lines are useless, will remove

**Comment by luigi-calori** [comment]
Addressed some comments, Now switch to  server side deployment,
 hope being able to test functionality today

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 11]
```diff
@@ -50,12 +66,67 @@ class Mesa(AutotoolsPackage):
...
+            base = Mesa.base_url + "/older-versions/%s.x" % (version.up_to(1))
+        else:
+            base = Mesa.base_url
+        if version > Version('10.4'):
+            name = 'mesa'
+        else:
+            name = 'MesaLib'
+        return base + "/%s/%s-%s.tar.gz" % (version, name, version)
+
+    def check_variants(self, var='', ver='', ref=''):
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: please use `conflicts('+galium', when='@:9.2.2')` statements as this is done during **conretization** instead of during the build.

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 12]
```diff
@@ -30,16 +30,32 @@ class Mesa(AutotoolsPackage):
...
+    base_url = "ftp://ftp.freedesktop.org/pub/mesa/"
+    url      = base_url + "12.0.3/mesa-12.0.3.tar.gz"
 
+    version('13.0.2', 'cb828f9f68e0cd6c7ef30b9c67aa7cf2')
+    version('13.0.1', '8415d4bb7837e6cfb0c819fdd19a643b')
+    version('13.0.0', '7205edb90d0396dc26d049fa495f6fd1')
     version('12.0.3', '60c5f9897ddc38b46f8144c7366e84ad')
 
+    variant('gallium', default=False, description="compile with gallium llvm sw rendering")
+    variant('drm', default=True, description="compile with drm")
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: this variant is not used

Comment (luigi-calori) [inline var/spack/repos/builtin/packages/mesa/package.py]: You are right, It was in the original version. I think that Mesa can be used also when a real GPU is present. Maybe there is a use case for that.
As soon as I manage to test on server, I' ll try to use it to reduce dependencies, like libdrm that now is a unconditional dep. I will test on server without in and then plan to leave it under drm variant.

Comment (davydden) [inline var/spack/repos/builtin/packages/mesa/package.py]: > I' ll try to use it to reduce dependencies, like libdrm that now is a unconditional dep. I will test on server without in and then plan to leave it under drm variant.

i don't know what `drm` is, but if it is indeed unconditional dependency, you should keep it as such regardless of whether or not it's provided by the system. If someone wants to build with system-provided `libdrm`, then they should configure external package in `packages.yaml`.

**Comment by mathstuf** [comment]
Is there any reasons we're bothering to add the 12 or 13 series? 17 is out…

**Comment by luigi-calori** [comment]
No reason apart the fact that when I started working on  that, I updated the spack package to the current latest and continued testing with that version.
I can try and see if the newest is working.
The strange thing is that it seem that mesa jumped from 13 series to 17 skipping all in-between
I'll try to test the 17 when possible

**Comment by mathstuf** [comment]
They are now using a year-based version scheme with quarterly releases. So 17.3 will probably be the last before 18.0 is released.

**Comment by davydden** [comment]
@luigi-calori since there has been some unification of mesa, can this PR be closed? Or rebase on top of develop?

**Comment by tgamblin** [comment]
@luigi-calori: closing b/c #5170 has been merged.  If you still need changes here, please rebase and reopen.
